src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,571 @@
+      'Push.ActivationStateMachine.handleEvent()',
+      'transition: ' + this.current.name + ' -(' + event.constructor.name + ')-> ' + maybeNext.name,
+    );
+    this.current = maybeNext;
+
+    while (this.pendingEvents.length > 0) {
+      const pending = this.pendingEvents[0];
+
+      this.rest.Logger.logAction(
+        this.rest.Logger.LOG_MAJOR,
+        'Push.ActivationStateMachine.handleEvent()',
+        'attempting to consume pending event: ' + pending.constructor.name,
+      );
+
+      maybeNext = this.current(this, pending);
+      if (!maybeNext) {
+        break;
+      }
+      this.pendingEvents.splice(0, 1);
+
+      this.rest.Logger.logAction(
+        this.rest.Logger.LOG_MAJOR,
+        'Push.ActivationStateMachine.handleEvent()',
+        'transition: ' + this.current.name + ' -(' + pending.constructor.name + ')-> ' + maybeNext.name,
+      );
+      this.current = maybeNext;
+    }
+
+    this.persist();
+    this.handling = false;
+  }
+}
+
+// Events
+export class CalledActivate {
+  constructor(machine: ActivationStateMachine, registerCallback?: RegisterCallback) {
+    if (registerCallback) {
+      machine.registerCallback = registerCallback;
+    }
+    machine.persist();
+  }
+}
+
+export class CalledDeactivate {
+  constructor(machine: ActivationStateMachine, deregisterCallback?: DeregisterCallback) {
+    machine.deregisterCallback = deregisterCallback;
+    machine.persist();
+  }
+}
+
+export class GotPushDeviceDetails {}
+
+export class GettingPushDeviceDetailsFailed {
+  reason: ErrorInfo;
+
+  constructor(reason: ErrorInfo) {
+    this.reason = reason;
+  }
+}
+
+class GotDeviceRegistration {
+  tokenDetails: API.TokenDetails;
+
+  constructor(deviceRegistration: DeviceRegistration) {
+    this.tokenDetails = deviceRegistration.deviceIdentityToken;
+  }
+}
+
+class GettingDeviceRegistrationFailed {
+  reason: ErrorInfo;
+  constructor(reason: ErrorInfo) {
+    this.reason = reason;
+  }
+}
+
+class RegistrationUpdated {}
+
+class UpdatingRegistrationFailed {
+  reason: ErrorInfo;
+
+  constructor(reason: ErrorInfo) {
+    this.reason = reason;
+  }
+}
+
+class Deregistered {}
+
+class DeregistrationFailed {
+  reason: ErrorInfo;
+  constructor(reason: ErrorInfo) {
+    this.reason = reason;
+  }
+}
+
+type ActivationEvent =
+  | CalledActivate
+  | CalledDeactivate
+  | GotPushDeviceDetails
+  | GettingPushDeviceDetailsFailed
+  | GotDeviceRegistration
+  | GettingDeviceRegistrationFailed
+  | RegistrationUpdated
+  | UpdatingRegistrationFailed
+  | Deregistered
+  | DeregistrationFailed;
+
+// States
+const NotActivated = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledDeactivate) {
+    machine.deactivatedCallback?.(null);
+    return NotActivated;
+  } else if (event instanceof CalledActivate) {
+    const device = machine.rest.device as LocalDevice;
+
+    if (device.deviceIdentityToken != null) {
+      // Already registered.
```

Comment:
According to the spec we should check `clientId` here [RSH3a2a1](https://sdk.ably.com/builds/ably/specification/main/features/#RSH3a2a1) and update token (not sure that it's valid for web push, but we have `WaitingForRegistrationUpdate` that makes me think it should be)